pkg/policy/manager.go

```diff
@@ -0,0 +1,455 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/cloudquery/cloudquery/internal/file"
+	"github.com/cloudquery/cloudquery/pkg/config"
+	"github.com/cloudquery/cloudquery/pkg/ui"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/hashicorp/go-version"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+const (
+	pathDelimiter    = "/"
+	versionDelimiter = "@"
+
+	cloudQueryOrg = "cloudquery"
+	gitHubUrl     = "https://github.com/"
+
+	defaultLocalSubPath   = ".cq/policy/"
+	defaultPolicyFileName = "policy"
+)
+
+var defaultSupportedPolicyExtensions = []string{"hcl", "json"}
+
+// ManagerImpl is the manager implementation struct.
+type ManagerImpl struct {
+	// policyDirectory points to the local policy directory
+	policyDirectory string
+
+	// Instance of a database connection pool
+	pool *pgxpool.Pool
+}
+
+var (
+	managerInstance *ManagerImpl
+	once            sync.Once
+)
+
+// Policy represents a single policy.
+type Policy struct {
+	// Organization is the policy org.
+	Organization string
+
+	// Repository is the policy repository.
+	Repository string
+
+	// RepositoryPath is the policy repository internal path.
+	RepositoryPath string
+
+	// Version is the git repository tag that should be used.
+	Version string
+
+	// SubPath is the policy sub-path.
+	SubPath string
+}
+
+// Manager is the interface that describes the interaction with the policy hub.
+// Implemented by ManagerImpl.
+type Manager interface {
+	// ParsePolicyHubPath parses a given policy hub path and returns a Policy object.
+	ParsePolicyHubPath(args []string, subPolicyPath string) (*Policy, error)
+
+	// DownloadPolicy downloads the given policy.
+	DownloadPolicy(ctx context.Context, p *Policy) error
+
+	// RunPolicy runs the given policy.
+	RunPolicy(ctx context.Context, execRequest *ExecuteRequest) (*ExecutionResult, error)
+}
+
+// NewManager returns the manager instance.
+func NewManager(policyDir string, pool *pgxpool.Pool) Manager {
+	// Singleton instantiation
+	once.Do(func() {
+		managerInstance = &ManagerImpl{
+			policyDirectory: policyDir,
+			pool:            pool,
+		}
+	})
+	return managerInstance
+}
+
+// ParsePolicyHubPath parses and validates the given arguments into the Policy struct.
+// Given args should follow the following semantic structure:
+// [(organization/)repository-name(@tag)] ([repository-path])
+func (m *ManagerImpl) ParsePolicyHubPath(args []string, subPolicyPath string) (*Policy, error) {
+	// Make sure the mandatory args are given
+	if len(args) < 1 {
+		return nil, fmt.Errorf("invalid policy path. Repository name is required but got %#v", args)
+	}
+	policy := &Policy{
+		SubPath: subPolicyPath,
+	}
+
+	// Parse and validate org/repository
+	orgRepoSplit := strings.Split(args[0], pathDelimiter)
+
+	// Parse org/repo
+	switch len(orgRepoSplit) {
```

Comment:
maybe use [this](https://github.com/cloudquery/cloudquery/blob/21d9b753117aa1bfecbf730b7065f8f32165558e/pkg/plugin/registry/organization.go#L15) function can be upgraded, seems like similar solutions no? or it can be used for the first part

Reply:
The error that gets returned by that function doesn't fit the use-case here. We could refactor this but I wonder if it is worth the effort since we're talking here about a few lines of code? WDYT?